File: vibeCoding/backend/python/fastapi-module/fastapi-agent-module-skill/references/tools/org_tools.py
# 组织架构 Tools（依赖 auth 模块）
# ✅ 修复 P0-A3: 所有 Tool 注入 current_user_id，用于租户隔离 + 审计

from src.agent.tools import tool


@tool(name="get_org_tree", description="获取组织架构树。查询当前用户所属租户下的完整部门组织结构")
async def get_org_tree(current_user_id: int) -> dict:
    """
    查询组织架构树（仅当前用户所属租户）

    Args:
        current_user_id: 当前登录用户ID（由系统自动注入）

    Returns:
        组织树
    """
    from src.auth.services.org_service import OrgService
    from src.auth.services.user_service import UserService

    # ✅ 修复 P0-A3: 通过 user 获取组织（多租户隔离）
    user = await UserService.get_user(current_user_id)
    if not user or not user.tenant_id:
        return {"error": "无法识别用户租户"}

    tree = await OrgService.get_org_tree(tenant_id=user.tenant_id)
    return {"org_tree": tree, "tenant_id": user.tenant_id}


@tool(name="get_org_detail", description="获取部门详情。根据部门ID查询详细信息（仅限当前租户）")
async def get_org_detail(org_id: int, current_user_id: int) -> dict:
    """
    查询部门详情

    Args:
        org_id: 部门ID
        current_user_id: 当前登录用户ID（由系统自动注入）

    Returns:
        部门信息
    """
    from src.auth.services.org_service import OrgService
    from src.auth.services.user_service import UserService

    # ✅ 修复 P0-A3: 校验部门归属当前用户租户
    user = await UserService.get_user(current_user_id)
    if not user or not user.tenant_id:
        return {"error": "无法识别用户租户"}

    org = await OrgService.get_org(org_id, tenant_id=user.tenant_id)
    if not org:
        return {"error": "部门不存在或无权访问"}

    return org


@tool(name="get_post_list", description="获取岗位列表。查询当前租户下的所有岗位")
async def get_post_list(current_user_id: int, status: int = 1, page: int = 1, page_size: int = 20) -> dict:
    """
    查询岗位列表（仅当前租户 + 分页）

    Args:
        current_user_id: 当前登录用户ID（由系统自动注入）
        status: 岗位状态，默认查询在职
        page: 页码
        page_size: 每页数量（最大50）

    Returns:
        岗位列表
    """
    from src.auth.services.post_service import PostService
    from src.auth.services.user_service import UserService

    # ✅ 修复 P0-A3: 多租户隔离 + 分页
    user = await UserService.get_user(current_user_id)
    if not user or not user.tenant_id:
        return {"error": "无法识别用户租户"}

    # ✅ 修复 P0-S6: 边界校验
    safe_page = max(page, 1)
    safe_page_size = min(max(page_size, 1), 50)

    posts = await PostService.list_posts(
        tenant_id=user.tenant_id,
        status=status,
        page=safe_page,
        page_size=safe_page_size
    )
    return {"posts": posts.items, "total": posts.total, "page": safe_page}


@tool(name="get_tenant_info", description="获取当前用户所属租户信息。仅返回当前用户租户，不暴露其他租户")
async def get_tenant_info(current_user_id: int) -> dict:
    """
    查询当前用户所属租户信息

    Args:
        current_user_id: 当前登录用户ID（由系统自动注入）

    Returns:
        当前租户信息
    """
    from src.auth.services.tenant_service import TenantService
    from src.auth.services.user_service import UserService

    # ✅ 修复 P0-A3: 仅返回当前用户租户，不再返回所有租户列表
    user = await UserService.get_user(current_user_id)
    if not user or not user.tenant_id:
        return {"error": "无法识别用户租户"}

    tenant = await TenantService.get_tenant(user.tenant_id)
    return {"tenant": tenant}

File: vibeCoding/backend/java/springboot-module/springboot-agent-module-skill/references/org-tools.java
package {basePackage}.agent.tool;

import {basePackage}.auth.entity.SysDept;
import {basePackage}.auth.entity.SysPost;
import {basePackage}.auth.entity.SysUser;
import {basePackage}.auth.service.SysDeptService;
import {basePackage}.auth.service.SysPostService;
import {basePackage}.auth.service.SysTenantService;
import {basePackage}.auth.service.SysUserService;
import jakarta.validation.constraints.Max;
import jakarta.validation.constraints.Min;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.Map;

/**
 * 组织架构 Tool
 * 全部按 tenant 过滤，多租户安全
 */
@Slf4j
@Component
public class OrgTools extends BaseTool {

    @Autowired
    private SysDeptService deptService;

    @Autowired
    private SysPostService postService;

    @Autowired
    private SysTenantService tenantService;

    @Autowired
    private SysUserService userService;

    /**
     * 获取当前租户组织架构树
     * 通过 userId 查询 tenantId，按 tenant 过滤
     */
    @AgentTool(name = "getOrgTree", description = "获取当前租户组织架构树")
    public List<Map<String, Object>> getOrgTree(
            @ToolParam(description = "当前用户ID（系统注入）") Long userId) {
        Long tenantId = getTenantId(userId);
        List<SysDept> depts = deptService.selectDeptListByTenantId(tenantId);
        return buildDeptTree(depts, 0L);
    }

    /**
     * 获取当前租户部门详情
     */
    @AgentTool(name = "getOrgDetail", description = "获取当前租户部门详情")
    public Map<String, Object> getOrgDetail(
            @ToolParam(description = "部门ID") Long orgId,
            @ToolParam(description = "当前用户ID（系统注入）") Long userId) {
        Long tenantId = getTenantId(userId);
        SysDept dept = deptService.selectDeptById(orgId);
        if (dept == null || !tenantId.equals(dept.getTenantId())) {
            return Map.of("error", "部门不存在或无权访问");
        }
        return Map.of(
            "deptId", dept.getDeptId(),
            "deptName", dept.getDeptName(),
            "leader", dept.getLeader() != null ? dept.getLeader() : "",
            "parentId", dept.getParentId(),
            "status", dept.getStatus()
        );
    }

    /**
     * 获取当前租户岗位列表（分页）
     */
    @AgentTool(name = "getPostList", description = "获取当前租户岗位列表")
    public List<Map<String, Object>> getPostList(
            @ToolParam(description = "当前用户ID（系统注入）") Long userId,
            @ToolParam(description = "状态（0正常 1停用）") @Min(0) @Max(1) Integer status,
            @ToolParam(description = "页码") @Min(1) Integer page,
            @ToolParam(description = "每页数量") @Min(1) @Max(50) Integer pageSize) {
        Long tenantId = getTenantId(userId);
        List<SysPost> posts = postService.selectPostListByTenantId(tenantId, status, page, pageSize);
        return posts.stream()
            .map(p -> Map.<String, Object>of(
                "postId", p.getPostId(),
                "postName", p.getPostName(),
                "postCode", p.getPostCode(),
                "status", p.getStatus()
            ))
            .toList();
    }

    /**
     * 获取当前用户所属租户信息
     */
    @AgentTool(name = "getTenantInfo", description = "获取当前用户所属租户信息")
    public Map<String, Object> getTenantInfo(
            @ToolParam(description = "当前用户ID（系统注入）") Long userId) {
        SysUser user = userService.selectUserById(userId);
        if (user == null) {
            return Map.of("error", "用户不存在");
        }
        Long tenantId = user.getTenantId();
        if (tenantId == null) {
            return Map.of("error", "用户未分配租户");
        }
        var tenant = tenantService.selectTenantById(tenantId);
        if (tenant == null) {
            return Map.of("error", "租户不存在");
        }
        return Map.of(
            "tenantId", tenant.getTenantId(),
            "tenantName", tenant.getTenantName(),
            "contactName", tenant.getContactName() != null ? tenant.getContactName() : "",
            "status", tenant.getStatus()
        );
    }

    /**
     * 通过用户ID获取租户ID
     */
    private Long getTenantId(Long userId) {
        SysUser user = userService.selectUserById(userId);
        if (user == null || user.getTenantId() == null) {
            throw new RuntimeException("用户未分配租户");
        }
        return user.getTenantId();
    }

    /**
     * 构建部门树
     */
    private List<Map<String, Object>> buildDeptTree(List<SysDept> depts, Long parentId) {
        return depts.stream()
            .filter(d -> parentId.equals(d.getParentId()))
            .map(d -> {
                Map<String, Object> node = new java.util.HashMap<>();
                node.put("deptId", d.getDeptId());
                node.put("deptName", d.getDeptName());
                node.put("children", buildDeptTree(depts, d.getDeptId()));
                return node;
            })
            .toList();
    }
}
